python/nano/test/pytorch/tests/test_backprop.py

```diff
@@ -0,0 +1,113 @@
+from bigdl.nano.pytorch.vision.models import vision
+from bigdl.nano.pytorch.algorithms.selective_backprop import SelectiveBackprop
+
+num_classes = 10
+batch_size = 256
+num_workers = 0
+data_dir = os.path.join(os.path.dirname(__file__), "../data")
+
+
+class CheckBatchSize(Callback):
+    def __init__(
+        self,
+        start: float = 0.5,
+        keep: float = 0.5,
+        end: float = 0.9,
+        interrupt: int = 2,
+        batch_size: int = 256
+    ):
+        self.start = start
+        self.keep = keep
+        self.end = end
+        self.interrupt = interrupt
+        self.batch_size = batch_size
+
+    def __should_selective_backprop(
+        self,
+        current_duration: float,
+        batch_idx: int,
+        start: float = 0.5,
+        end: float = 0.9,
+        interrupt: int = 2,
+    ) -> bool:
+        is_interval = ((current_duration >= start) and (current_duration < end))
+        is_step = ((interrupt == 0) or ((batch_idx + 1) % interrupt != 0))
+
+        return is_interval and is_step
+
+    def on_train_batch_end(self, trainer, pl_module, outputs, batch: Any, batch_idx: int, dataloader_idx: int) -> None:
+        elapsed_duration = float(trainer.current_epoch) / \
+            float(trainer.max_epochs)
+        if self.__should_selective_backprop(elapsed_duration, batch_idx, self.start, self.end, self.interrupt):
+            current_batch_size = len(batch[1])
+            ideal_batch_size = int(self.keep * self.batch_size)
+            assert current_batch_size == ideal_batch_size, \
+                'Batch size is not right.'
+
+
+class ResNet18(nn.Module):
+    def __init__(self, pretrained=True, include_top=False, freeze=True):
+        super().__init__()
+        backbone = vision.resnet18(pretrained=pretrained, include_top=include_top, freeze=freeze)
+        output_size = backbone.get_output_size()
+        head = nn.Linear(output_size, num_classes)
+        self.model = torch.nn.Sequential(backbone, head)
+
+    def forward(self, x):
+        return self.model(x)
+
+
+model = ResNet18(pretrained=False, include_top=False, freeze=True)
+loss = nn.CrossEntropyLoss()
+optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
+
+
+class TestLightningModuleFromTorch(TestCase):
+
+    def test_selective_backprop(self):
+        pl_model = LightningModuleFromTorch(
+            model, loss, optimizer,
+            metrics=[torchmetrics.F1(num_classes), torchmetrics.Accuracy(num_classes=10)]
+        )
+        data_loader = create_data_loader(data_dir, batch_size, num_workers, data_transform)
+        loss_fn = nn.CrossEntropyLoss(reduction='none')
+        sb = SelectiveBackprop(start=0.25, keep=0.53, end=0.75,
+                               scale_factor=0.1, interrupt=2, loss_fn=loss_fn)
+        batch_size_check = CheckBatchSize(start=0.25, keep=0.53, end=0.75, interrupt=2, batch_size=batch_size)
+        trainer = Trainer(max_epochs=8, log_every_n_steps=1, callbacks=[sb, batch_size_check])
```

Comment:
How long does it take to run this test? Can we make `max_epochs` smaller?

Reply:
Yes we can. I think it can be reduced to 4.